UI: Combined source context menu and media controls

cg2121 opened the pull request:
### Description
Combination of #2380 and #1838 (by @DDRBoxman).

### Motivation and Context
Mostly to test the combined PRs since they rely on each other. Will need a 'Seeking Testers' flag.

### How Has This Been Tested?
Selected sources with and without media to make sure everything ran as expected.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### Test build links
Please leave your feedback below after testing!

Windows, macOS, and Linux: https://github.com/obsproject/obs-studio/actions/runs/101066040

xiCO2k commented:
I will be available to test, any one can create a build with it?

Fenrirthviti commented:
Builds for testing:

[Windows](https://dev.azure.com/obsjim/1437b187-9c54-400d-a378-3dff67a18804/_apis/build/builds/3565/artifacts?artifactName=winbuild&api-version=5.1&%24format=zip)
[macOS](https://dev.azure.com/obsjim/1437b187-9c54-400d-a378-3dff67a18804/_apis/build/builds/3565/artifacts?artifactName=macbuild&api-version=5.1&%24format=zip)

xiCO2k commented:
Awesome will test on a livestream tonight, I will have 3 scenes with videos. and will give you guys some feedback

xiCO2k commented:
![image](https://user-images.githubusercontent.com/823088/81431372-ef876980-9158-11ea-9705-00aeed2dc62e.png)
Found this small icon overlapping case

dodgepong commented:
That may be more of an issue with the context menu itself, so I would report that issue here: https://github.com/obsproject/obs-studio/pull/1838

Feedback on this PR should be strictly about media controls.

xiCO2k commented:
Sounds good will do

cg2121 commented:
I've now added the ability to set the timer to count down. This is accessible by clicking the properties button in the media controls.

I've also added hotkeys for the media controls, so users don't have to set hotkeys for each media source.

dodgepong commented:
Looking good so far! Really glad to see generic media control hotkeys added in this iteration.

Some feedback on the controls themselves:

* The control buttons need tooltips
* The Previous button should be the left-most button (Next should be right-most, as it is now)
* I can't help but feel like Restart needs to be its own button. Sometimes you want to restart in the middle of playback, or while playback is paused. Restart is the "meme" button: Imagine someone playing an airhorn sound over and over -- they would do this by repeatedly pressing the Restart button, usually before playback of the sound ends.
    * Thus, the order of buttons should be [Prev] [Play/Pause] [Stop] [Restart] [Next]

For some of the rest:

* I'm not crazy about the cogwheel to enable the countdown timer, and honestly would rather see it behind a (sigh) right-click menu if anything, or in setting somewhere else. Also I would like to hear an argument as to why countdown shouldn't be the default (I think it should be the default).
* If playback is paused, checking the Countdown timer option does not toggle the display until playback resumes
* When dragging the timeline slider while playback is paused, when I let go of the slider, the current frame does not update until playback resumes. I can't help but feel like people would like the current paused frame to be updated to the new position even while paused. I can understand if this might be difficult, though.
* The toggle on the Play/Pause hotkey seemed like it wasn't working properly for me. I set both Play and Pause to Space, expecting that I could hit space to play when paused/stopped or pause when playing. However, it seemed to trigger both play and pause at the same time, which lead to playback either not starting at all, or stopping and then immediately starting again.
* I found it weird that the media controls hotkeys were not grouped together in the Hotkeys menu -- is that due to a weird way that they are sorted?

xiCO2k commented:
Looking go also here. Used it on my live stream yesterday and worked out really good.

tc-vf commented:
Thanks so much everyone, this is a *massive* step forward in my deployment.  I tested on a live stream today, although using the build downloaded from above (ie. without the countdown option). I'm struggling with building my own - will work on that...

Some feedback:

- 100% agree re time-to-go being better than a plain time-played-so-far display (for me, anyway). Would a global toggle setting for that in the main preferences be possible?

- Is it planned to implement the dock UI as well as the context bar, as per previous discussion in #2380 ?  Any timer is a giant step forward, but also having the program controls, or at least timer(s) next to the program audio meter would be so much better.  I understand the issues with multiple sources per scene, but I think I could live with it - it works fine for audio.  Or, could there be a single "Time to end of all scenes" timer?

- I agree that the source should refresh when dragging the slider while paused.

cg2121 commented:
> The control buttons need tool tips

Done.


> The Previous button should be the left-most button (Next should be right-most, as it is now)

Most media players organize the buttons the way I have it.


> I can't help but feel like Restart needs to be its own button. Sometimes you want to restart in the middle of playback, or while playback is paused. Restart is the "meme" button: Imagine someone playing an airhorn sound over and over -- they would do this by repeatedly pressing the Restart button, usually before playback of the sound ends. 

I feel like this would be too much noise and clutter.


> I'm not crazy about the cogwheel to enable the countdown timer, and honestly would rather see it behind a (sigh) right-click menu if anything, or in setting somewhere else. Also I would like to hear an argument as to why countdown shouldn't be the default (I think it should be the default).

The cogwheel could be used for other settings in the future. I've also set the countdown timer to be default now.


> If playback is paused, checking the Countdown timer option does not toggle the display until playback resumes

Fixed.


> When dragging the timeline slider while playback is paused, when I let go of the slider, the current frame does not update until playback resumes. I can't help but feel like people would like the current paused frame to be updated to the new position even while paused. I can understand if this might be difficult, though.

Media source internals would need to be changed for this to work.


> The toggle on the Play/Pause hotkey seemed like it wasn't working properly for me. I set both Play and Pause to Space, expecting that I could hit space to play when paused/stopped or pause when playing. However, it seemed to trigger both play and pause at the same time, which lead to playback either not starting at all, or stopping and then immediately starting again.

Since the space bar is a QT shortcut for triggering the current button, the space bar wouldn't work here.


> I found it weird that the media controls hotkeys were not grouped together in the Hotkeys menu -- is that due to a weird way that they are sorted?

Fixed.

psxde commented:
Thank you very much for this pull request. I started to use the test build linked above as media player for live shows. First: it works great!

The only thing which I miss a lot - and already mentioned - is an overview of the current playing media sources without clicking on the source, especially in the studio mode in live situation. Maybe they could be listed in a separate dock exactly like the audio mixer. At least I would need the remaining time of the media sources.

xiCO2k commented:
@psxde That would be great, probably on the next version :)

psxde commented:
Workaround is possible with a custom html gui and the websocket plugin (see here: https://github.com/Palakis/obs-websocket/pull/497)

tconley commented:
In the meantime, is there a version of this that's available as a plugin? Is there a latest build we can test?

tconley commented:
> Workaround is possible with a custom html gui and the websocket plugin (see here: [Palakis/obs-websocket#497](https://github.com/Palakis/obs-websocket/pull/497))

Are you working on this approach? I'd love to help.

dodgepong commented:
The latest builds of this PR (as of this posting) can be found here: https://github.com/obsproject/obs-studio/actions/runs/101066040

tt2468 commented:
@psxde That's a good idea to add as a request to my obs-websocket pr. I'll work on it.

adamatos commented:
> Thank you very much for this pull request. I started to use the test build linked above as media player for live shows. First: it works great!
> 
> The only thing which I miss a lot - and already mentioned - is an overview of the current playing media sources without clicking on the source, especially in the studio mode in live situation. Maybe they could be listed in a separate dock exactly like the audio mixer. At least I would need the remaining time of the media sources.

I couldn't agree more. This is a HUGE improvement in being able to time media playback when live, just unfortunate it kinda falls apart when using it in studio mode.

psxde commented:
@tconley @tt2468 I have used the currently test version of this pr and the changes in obs-websocket and iterate through every active source to get currently playing media and built a simple html gui, so it's already possible with this workaround (thanks, it helps A LOT!). I would love to help integrate this better 'native' as soon as time and knowledge is sufficient.

tconley commented:
@psxde  - are you in the obs discord? would love to chat with your about your interface. I made a simple html gui with play/pause buttons that talks to @tt2468 's additions to websocket, but have yet to impliment a draggable slider bar and countdown timer for playback and other key features. lets' collab

jgclark commented:
Great work, thanks.
I've tested, and I'm with @psxde that it would be great to have the media source automatically  selected -- at least where there's only a single relevant media source in the scene.

juanliempo commented:
I recently updated to 25.0.8 on win64, I wanted to test this build (2020-05-11-ab7e449e-25.0.7-win64.zip) since I am needing media/vlc controls. Am I just copying all the files over or is there something else I should be doing instead? Still somewhat new, thanks in advance for directions yall can provide.

Fenrirthviti commented:
Grab the test build and extract it to its own folder somewhere. I would not recommend overwriting your main install of OBS with this test build, as it may be unstable.

psxde commented:
@tconley sorry, currently I am not using discord. I have a very special interface for a moderator view; currently without any controls (but play/pause media is planned and should be easy to implement). Currently, time is rare but I try to checkin the current state the next days into github. 

In short I currently iterate through the sources on start and when the scene changes and check for type = "ffmpeg_source", "vlc_source", then place them into the webgui and request the duration and state. While a source is playing (MediaStarted, MediaPlaying) I set a timer to get the current time and subtract it from the duration and format it (currently only a string, not a slider)

tt2468 commented:
@psxde Are you doing this on obs-websocket or the obs api? On obs-websocket, I added a request specifically for getting the status of all media players. On the OBS c api, you can use `obs_enum_sources`

KonajuGames commented:
The time remaining counter works really well, except for studio mode. The timers should work on the media in the Program output, not the Preview.  In its current state, this is unfortunately not useful for broadcast production use.

Claudio3112 commented:
> Prendi la build di test ed estraila nella sua cartella da qualche parte. Non consiglierei di sovrascrivere la tua installazione principale di OBS con questa build di test, poiché potrebbe essere instabile. Obs64.exe può quindi essere eseguito dalla directory bin \ 64bit.

Scusa io non riesco avviare applicazione test, ho provato ma mi da prima un errore che manca libreria vcruntime140_1.dll poi impossibile avviare correttamente applicazione.

jgclark commented:
> Hotkeys for media control are already available and working in OBS, you can find them in Settings -> Hotkeys.

My comment made clear that I'm aware that they exist. But I'm reporting that they're not working for me as I expect in this build. Please educate me whether my expectation is right that they should work on 'Media Source' items.

Fenrirthviti commented:
It is a little confusing, you have to find the hotkeys for the specific source in question (scroll down in the list and look for the specific media source you want to set a hotkey for). There is no global hotkey for all media sources.

jgclark commented:
> It is a little confusing, you have to find the hotkeys for the specific source in question (scroll down in the list and look for the specific media source you want to set a hotkey for). There is no global hotkey for all media sources.

Thanks for your help. What's particularly confusing is that in this build (but not in 25.0.8) I find "Play Media", "Pause Media", "Restart Media", and "Stop Media" almost at the top of the list of possible Hotkeys -- i.e. before it starts listing settings for particular sources or scenes. What are these supposed to do if not general keys for any source?

dodgepong commented:
They are supposed to work for the currently-selected source.

Can you give more details about what exactly isn't working? What hotkeys are you trying to set, and what keys are you setting them to? What happens when you press the hotkeys?

CodeYan01 commented:
Hi, I recently tried the test build for Windows. Are playlists supported here? I tried using a .m3u8 playlist but it doesn't seem to work. It would be great to have a playlist feature.

jgclark commented:
> They are supposed to work for the currently-selected source.
> 
> Can you give more details about what exactly isn't working? What hotkeys are you trying to set, and what keys are you setting them to? What happens when you press the hotkeys?

Thanks, @dodgepong. I'm trying to set hotkeys for "Play Media", "Pause Media", "Restart Media", and "Stop Media" almost at the top of the list of possible Hotkeys, with single key definitions (R, S, P etc.) as well as multi-key (↑⌘R, ↑⌘S, ↑⌘P etc.).  I've tried it on the equivalent for the video file itself, that's the only source in a scene. I've tried keys with the source actively highlighted in the scene list or not.  Nothing at all happens that I can detect. 

Running that build on latest version of macOS Catalina.

cg2121 commented:
I've now replaced the global hotkeys for the current media with shortcuts. This simplifies the code and doesn't clog up hotkeys UI.

Shortcuts:
Space - play/pause
S - stop
R - restart
P - playlist previous
N - playlist next

unknowndomain commented:
Just been playing around with the latest version. (I think, so many builds there!)

It's fantastic to see this, it would have been so useful to have this today.

I have a few ideas/suggestions/minor "bugs" in the GUI on macOS catalina

# Bug: Narrow window
At it's narrowest the GUI just doesn't make sense in OBS, there should probably be some graceful scaling when the window is narrow, i.e turning off the words filter and properties which are a massive waste of space. 

Removing the playbar, and maybe making the timer clickable so it switches from progress to remaining? Also some overlapping icon issues.

<img width="832" alt="Screenshot 2020-06-18 at 01 34 50" src="https://user-images.githubusercontent.com/147143/84964615-76d6de00-b104-11ea-8f87-2b39edb16ae3.png">

# Bug: Auto width timer issue
As the time changes the playbar length changes by a few pixels. This is easier to see when you zoom right in.

# Suggestion: Time positions
A common UX pattern is to put the time in on the left and the time remaining on the right...

You can see that here in QuickTime but it exists in many places and helps to clarify the meaning of the information as both then describe the time on the relivent side of the puck.

If you did this could could omit the minus on the right side too. And on that topic perhaps only showing hours if the video is that long would be helpful.

<img width="472" alt="Screenshot 2020-06-18 at 01 43 04" src="https://user-images.githubusercontent.com/147143/84964826-0ed4c780-b105-11ea-9504-d85c1180d002.png">

# Suggestion: Studio mdoe
On larger screens when using studio mode as someone else has said it should alight the media controls to the left edge of the program output, and the context menu controls to the preview output.

# Request: Filter options
It might be helpful for discoverability to show the filter options list on a button that either drops down or cycles through the available options.

# Suggestion: Scaling
The two scaling buttons at present are super helpful, but I fear for the volume of feature requests in future where people say...

I clicked the button but it stretched the video, and vica versa why are there black borders on my video. 

I'd suggest these icons need more work, or at least for LTR readers that the best option (borders) be the left most one.

# Bug: Playlist controls
The playlist controls are not very intuitive in my view, I am used to using a separate scene for each video and so having a forward/backward button was confusing, I expected that the back button would restart the video and it didn't do anything nor did forward. I would suggest that when there is only one video that these be disabled and greyed out to indicate that, and/or that the back button restart the current video.

This might be a useful setting to have it act like a CD player where back starts again, and double click goes to the previous video. Either configurable or as an Alt + Click type option.

# Request: Play button shape
I don't know if it's my eyes but the play button feels slightly stretched width wise, and I feel like it should be an equalateral triangle like the playlist controls.

jgclark commented:
Excellent comments from @unknowndomain, which echo many things I was going to suggest, particularly relating to how it works on a small screen size.

The only place I differ is on "I would suggest that when there is only one video that these be disabled and greyed out to indicate that ..."  I'd suggest that where there is only one video in a scene that it is auto-selected.

dodgepong commented:
> I'd suggest that where there is only one video in a scene that it is auto-selected.

Strong disagree on this one. The entire point of the bar "context menu" is to give extra options related to the currently-selected source. If the currently-selected source does not need media controls, then no media controls should be visible.

I would prefer an optional dock that shows controls for all media sources in the currently-selected scene as a solution to that problem.

jgclark commented:
> > I'd suggest that where there is only one video in a scene that it is auto-selected.
> 
> Strong disagree on this one. The entire point of the bar "context menu" is to give extra options related to the currently-selected source. If the currently-selected source does not need media controls, then no media controls should be visible.

Eh? I didn't suggest that if a source doesn't need media controls that it should have some. That would indeed be daft. Rather, that if a scene only has one source that should have media controls, then it should be auto-selected to save time. There would be no confusion as to what was being controlled.

unknowndomain commented:
@jgclark I think the point @dodgepong is making though is that the bar isn't related to studio mode, i.e. the buttons under preview and program are not linked for all users, so in live punching mode the left hand scene name indicates the selected scene.

Although this isn't a sword I'd want to die on, I don't personally think there is a need to use that much visual space to repeat information as it's already highlighted in the selected source below, further more if the other PR for a DSK (Global Scenes) was implemented a issue was raised that this too should only have one item selected at a time. Effectively this means the highlighted source in either regular or global scenes would already be highlighted blue so this information is duplication on valuable screenspace.

BenSa0112 commented:
Hey, just a simple question. Is it possible to make this PR as a Plugin? And will it be added to Master-branch?

Fenrirthviti commented:
This makes core changes to OBS, it would not be possible to make as a plugin. When all the testing has completed, it will be merged to the repository.

koerb commented:
great work!
if there would be a possibility to show the playlist as a dock or sth would be awesame

unknowndomain commented:
Haha, @koerb we're almost back full circle to where this idea originally came from as a dock with play controls :D

adamatos commented:
Its funny. Early on the comments reflected the need for it to be a dock as
it was the only way for it to work in various work flows. Unfortunately now
we have a lot of visual clutter and a tool not useful except for the most
of use cases.

On Fri, Jun 19, 2020 at 1:08 PM Tom Lynch <notifications@github.com> wrote:

> Haha, @koerb <https://github.com/koerb> we're almost back full circle to
> where this idea originally came from as a dock with play controls :D
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/obsproject/obs-studio/pull/2886#issuecomment-646798630>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARND4NXZ74TIPLTSX6DWQ3RXOSS7ANCNFSM4M326GCQ>
> .
>
-- 
- Adam Matos

unknowndomain commented:
@adamatos Would it therefore not make sense to do both...

Simple controls like play, pause, restart, next and previous with a sime remaining counter on the master output.

And a Media Control Dock with the full controls at the top something like this:

<img width="377" alt="Screenshot 2020-06-19 at 19 26 31" src="https://user-images.githubusercontent.com/147143/85168842-d64b0000-b262-11ea-8c45-27714ab7ff81.png">

Below would be a list of items like the editable playlist in source properties, but with an icon showing which is playing or a button to skip to it.

adamatos commented:
Again. This is just my personal opinion, but the UI conventions that are
established in the way the audio controls work should have been maintained
with these. A set of buttons under the windows that does or does not work
depending on the scene is less then ideal and rip to create confusion. It
also completely ignores the concept that users might have more then one
element playing in a single scene (Logo animation, background, video asset)
All at once. Something a dock would be able to easily address.

In regards to both.... more UI options are rarely more options with regard
UX. Its often just an opportunity for more confusion.

On Fri, Jun 19, 2020 at 1:28 PM Tom Lynch <notifications@github.com> wrote:

> @adamatos <https://github.com/adamatos> Would it therefore not make sense
> to do both...
>
> Simple controls like play, pause, restart, next and previous with a sime
> remaining counter on the master output.
>
> And a Media Control Dock with the full controls at the top something like
> this:
>
> [image: Screenshot 2020-06-19 at 19 26 31]
> <https://user-images.githubusercontent.com/147143/85168842-d64b0000-b262-11ea-8c45-27714ab7ff81.png>
>
> Below would be a list of items like the editable playlist in source
> properties, but with an icon showing which is playing or a button to skip
> to it.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/obsproject/obs-studio/pull/2886#issuecomment-646806796>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARND4NIBD2AC4YZ7U2IS6TRXOU5NANCNFSM4M326GCQ>
> .
>
-- 
- Adam Matos

renerlemes commented:
Used it on my live stream by 6 times and worked out really good. I missed the dock being always visible.

Fenrirthviti commented:
I am also in favor of providing a dock with all the media controls listed there, but the intent of the specific addition to the context menu is to keep UI clutter down for the majority of users who won't need to see the status of every media source at once.

I think having an optional dock to enable is probably a good move, but leaving it off by default. The original PR had a media controls dock, if I remember correctly, and I support it being re-added.

dodgepong commented:
The original PR put a single media control in a dock. It was not a list of all media controls in the scene, only a dedicated dock where media controls for the selected media source would go. The functionality in this PR duplicates that functionality. I don't believe there has ever been a version that had a dock full of controls for all media sources in a scene, though I agree that having it as an optional dock is not a bad idea in addition to this implementation.

MarioMey commented:
Sorry to write this here... but I don't know how to download this pull to compile and test/use it. If it is working well, I need it for my job!

Please, some who can point me?

Fenrirthviti commented:
Links to the test builds are at the bottom of the first post.

MarioMey commented:
> Links to the test builds are at the bottom of the first post.

Thanks for the response. I sayd that I downloaded that test build (2020-05-11-ab7e449e-25.0.7-linux64.tar.gz), installed it but I can't launch it: I don't find the executable. In synaptic, I can see  "obs-studio" as a installed packet, but there's no "obs" nor "obs-studio" command to launch it.

MarioMey commented:
I found executable (`/home/runner/work/obs-studio/obs-studio/obs-studio-portable/bin/64bit`), but it is asking for older libav* packets. I already created 3 symlinks to the newer versions... but it still asking for more. I don't know for how much time it will going on asking for them...

Fenrirthviti commented:
I would recommend just compiling yourself on Linux, the build we provide is only compatible with Ubuntu systems. If you need assistance building, this PR is not the right place to get help, please use the Discord or Forums. https://obsproject.com/help

foxafillion commented:
I have spent roughly 30 hours streaming with this version of OBS, the ability to see how much time is left on videos is extremely useful for running live events with some prerecorded stuff in between segments. I haven't really tested the buttons much but so far the software has been running without issue. I'm not sure if someone has mentioned this but the issue I'm not sure if there's an answer to is that I always run in studio mode because I need to see preview when running live shows. When I play a video, I can see how much time is left as long as I maintain both preview and live to be the same scene, if I play a video and preview the next scene I can no longer see how much time is remaining on the live video. It would be nice to be able to have on the right side a live timeline of the media playback and on the preview side have it's own separate thing or perhaps just show how long a source is. Is this something that's in the works? Or perhaps it's already there and I just don't know what I'm doing.

BTW thanks so much for your work on this, it seriously has made my life so much less stressful.

MarioMey commented:
> I have spent roughly 30 hours streaming with this version of OBS, the ability to see how much time is left on videos is extremely useful for running live events with some prerecorded stuff in between segments. I haven't really tested the buttons much but so far the software has been running without issue. I'm not sure if someone has mentioned this but the issue I'm not sure if there's an answer to is that I always run in studio mode because I need to see preview when running live shows. When I play a video, I can see how much time is left as long as I maintain both preview and live to be the same scene, if I play a video and preview the next scene I can no longer see how much time is remaining on the live video. It would be nice to be able to have on the right side a live timeline of the media playback and on the preview side have it's own separate thing or perhaps just show how long a source is. Is this something that's in the works? Or perhaps it's already there and I just don't know what I'm doing.
> 
> BTW thanks so much for your work on this, it seriously has made my life so much less stressful.

+1000 for this suggestion. I managed to compile this pull request and I love this feature. And I was thinking in another feature and I wanted to suggest somewhere, maybe this could be the place... maybe not. If not, please delete this comment and sorry.

I will do streaming for 16 days in a week. It's like a TV program, I'll be live and playing some videos. These videos were sent by very different people. Some videos start well, some others don't... they appear in the video touching the phone screen to start the record. As I will be the one who play that videos, I have to pre-crop every video at the start with a video editing software (end is no problem, I cut them with OBS before they end). In my case, @foxafillion, I can't use Studio Mode, because I will be live, acting.

Now, the suggestion: every Media Source could have In-Out marks. They could be set in the timeline and be saved as properties.

Fenrirthviti commented:
I would say at this point we should consider this feature complete, for the sake of getting this ever merged at all. I don't think anyone disagrees that there are a thousand and one little nice things that can be added to this feature, but if we keep focusing on what could be instead of what is, this will never be finished.

Let's focus on reviewing this PR as-is, and save discussion on the wish list until after it has been merged.

tconley commented:
can't wait to see this merged. i'll be using this and obs music edition as my main setup :-)

unknowndomain commented:
Feature complete yes but many issues have been raised about layout and design bugs which need to be fixed before it can be released.

> 
> On 14 Jul 2020, at 05:29, Joel Bethke <notifications@github.com> wrote:
> 
> ﻿
> I would say at this point we should consider this feature complete, for the sake of getting this ever merged at all. I don't think anyone disagrees that there are a thousand and one little nice things that can be added to this feature, but if we keep focusing on what could be instead of what is, this will never be finished.
> 
> Let's focus on reviewing this PR as-is, and save discussion on the wish list until after it has been merged.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Jonnycp commented:
There is an update for obs 25.0.8?

QNCthomas commented:
Just installed and testing the mac version.    In studio mode, the timer disappears when I click on other items to preview, which is our workflow.  Always preview before taking, and need the countdown to know when to take.   Even if in the next revision, that's fine, since this request is very long in the tooth.  Thank you all for your effort!

exeldro commented:
For anyone looking for a media controls dock, I made a plug-in based on #2380.
Source: 
https://github.com/exeldro/obs-media-controls
Plugin: https://obsproject.com/forum/resources/media-controls.1032/

jp9000 commented:
I want to pose an important question to everyone:

Are the buttons that are on the context bar by default the correct buttons to have? Right now the buttons are:

- Visibility Toggle -- Already in the sources list.  Is it useful to have it appear here redundantly?
- Selection Lock -- Already in the sources list.  Is it useful to have it appear here redundantly?
- Properties -- [x] This is good
- Filters -- [x] This is good
- Rotate Counter-clockwise 90° -- Is this something people need to do commonly?
- Rotate Clockwise 90° -- Is this something people need to do commonly?
- Flip Horizontally -- Is this something people need to do commonly?
- Flip Vertically -- Is this something people need to do commonly?
- Stretch to Screen (which ignores aspect ratio) -- Not sure how often users need to stretch ignoring aspect ratio, and probably needs better tooltip and context menu wording
- Fit to Screen (which repsects aspect ratio) -- [x] Probably good, but probably needs better tooltip and context menu wording
- Center -- [x] Probably good
- Open Transform Dialog -- Is this actually useful to always have access to from the toolbar?

I said this elsewhere but we need divergent thinking and should probably think about what to put here, then come to a concensus on what buttons are the most ideal.  In my opinion, anything displayed to users front and center needs to be catered for the novice computer user as well as the most common needs.

Additionally, having too many buttons or items could make the toolbar seem intimidating to those aforementioned type of users.  I'd prefer things to feel as friendly as possible, but that's definitely a subjective thing.

Again, the focus for this should really be users with novice computer knowledge, as they are the most common users. This shouldn't be a toolbar of convenience for advanced users, but provide advanced functionality in a way that even users of novice computer knowledge would appreciate and understand.

WizardCM commented:
My opinions on the buttons:
* Visibility/Lock toggles - I think it makes sense for them to be here, as it's about modifying the sceneitem, and the most obvious edit you can do is hide/lock it.
* Rotate - Personally, exposing rotation _at all_ is a good thing. I'd love handles on the sceneitem, but being able to quickly rotate things (eg. webcam borders) is useful.
* Flip - In the context of having reposition/resize/rotate, I think having flip is handy. Useful for webcams especially. Example use case: streamer with a green screen, wants to move webcam from one corner to another while still looking "towards the center". Drag, click to flip, done.
* Stretch to screen - after using this, I can't imagine a use for this. I'd be OK removing it
* Fit to screen - Very useful after adding a new item like an overlay or a temporary image. "Hey I saw this great picture earlier today - *drag and drop into OBS*, *Fit to screen* *done*"
* Open Transform Dialog - This could be either way for me, but it has its uses - especially when you're initially configuring your scenes, being able to do things pixel-by-pixel is a **godsend**, and having an easy way to learn about the dialog and be able to make quick changes, I believe it should stay.

In terms of Reset Transform, no thanks. It modifies too many things at once, and without undo could be very destructive.

I agree that the number of buttons right now should be the max.


@cg2121 I did find a bug with the media controls though.

Properties for a local file media source:
* Loop, Restart playback when source becomes active, Show nothing when playback ends
* Everything else disabled

1. Enable studio mode
2. wait 2 seconds
3. Disable studio mode
4. The media controls progress bar no longer functions. Restarting playback or switching to another scene is the only way to make it work again.

mufunyo commented:
My thoughts:

- Centre should probably be split into horizontal and vertical - if you want both you can just click one after the other.
- Stretch is unnecessary (fit should be the most used option, needing to stretch is rare).
- Reset is useful and I disagree with comments about accidentally clicking it.
- A button that opens the scale filtering menu could be useful.
- The mirror/flip buttons are useful, but I don't think rotate is very useful. #2212 would be a better fit for easy rotation, especially if holding shift while rotating the source will snap to 45/90/180 degree rotations.

jp9000 commented:
I fixed the issue with seek where holding the seek at the end position would cause playback to stop in 83913e1d31.  Additionally, I also made it so that it will actively show a preview of the seek position wherever you have the seek slider held to while seeking like a normal video player would.  Finally, I made the seek slider an absolute position slider, that way if you click an empty part it'll set the slider position to the clicked position rather than try to do a "page up" or "page down" like normal sliders.

I'm still undecided on the buttons to show on here.

WizardCM commented:
The seeking feels much better, nice work Jim.

I notice that the timer tooltip doesn't appear while dragging, but shows fine when hovering anywhere in the timeline. Not sure if Qt allows displaying the tooltip during dragging, but it would be a nice tweak.

![image](https://user-images.githubusercontent.com/941350/88152789-e1b98e80-cc47-11ea-8085-7deee412b6cd.png) ![image](https://user-images.githubusercontent.com/941350/88152859-fbf36c80-cc47-11ea-8c6e-f46d45cff943.png)


I also notice the timer to the right doesn't update when you reposition the slider while paused. Is that expected behaviour?
![image](https://user-images.githubusercontent.com/941350/88152825-f1d16e00-cc47-11ea-8be9-1c81bf985c2a.png)

Finally, the "right click to toggle total/remaining" feels very... off? I feel having that on left click would be fine. It's very undiscoverable though.

skeletonbow commented:
> Are the buttons that are on the context bar by default the correct buttons to have? Right now the buttons are:
>     * Visibility Toggle -- Already in the sources list.  Is it useful to have it appear here redundantly?
>     * Selection Lock -- Already in the sources list.  Is it useful to have it appear here redundantly?
>     * Properties -- [x] This is good
>     * Filters -- [x] This is good

Agreed.

>     * Rotate Counter-clockwise 90° -- Is this something people need to do commonly?
>     * Rotate Clockwise 90° -- Is this something people need to do commonly?
>     * Flip Horizontally -- Is this something people need to do commonly?
>     * Flip Vertically -- Is this something people need to do commonly?

Useful, but personally I'd prefer to have a Transform dock that includes a lot of the common transform options, which could be then stacked on other docks and brought to the front when scene editing.  Myself and a lot of others working on visual effects manipulate transforms a lot, and having to right click and go to the transform menu to access everything is a big dent in workflow, especially when manipulating transforms on a number of objects one after another.  Having that in a dock would mean it's active for the currently selected scene item which would be a huge workflow win.

Would it be possible in theory for such a feature to be made as a plugin, or would that need to be done in core?

>     * Stretch to Screen (which ignores aspect ratio) -- Not sure how often users need to stretch ignoring aspect ratio, and probably needs better tooltip and context menu wording
>     * Fit to Screen (which repsects aspect ratio) -- [x] Probably good, but probably needs better tooltip and context menu wording
>     * Center -- [x] Probably good

All nice conveniences when scene editing.  Could instead be in the above mentioned Transform Dock idea.

>     * Open Transform Dialog -- Is this actually useful to always have access to from the toolbar?  This is a very advanced dialog that is rarely needed by the vast majority of users in my humble opinion.

What about having the button bar configurable like many programs allow?  VLC player, most office suite apps, audacity, tonnes of others have a button bar with user configurable buttons.  Default installation includes only the most commonly used button features for the typical user, but allows customization by power users to include whatever buttons they find useful for their workflow.  I customize audacity and VLC player a fair bit in this regard.  Would be great to have similar functionality in OBS

> Also, what about a button for other features than those that novice users might appreciate having, such as, say, Reset Transform? Maybe it's not that useful for most novice users, I'm not sure.

I'd definitely not have a Reset Transform button as it is all to easy to accidentally click on one thing when you meant to click on the thing beside it.  Accidentally resetting a transform can potentially take a while to reconfigure it from scratch, especially without an undo/redo feature as someone mentioned above.  A reset button could be fine in a transform dock however as it isn't showing all the time, and would mimic some or all of the functionality of the transform dialog box, or even replace it.

> I said this elsewhere but we need divergent thinking and should probably think about what to put here, then come to a concensus on what buttons are the most ideal. In my opinion, anything displayed to users front and center needs to be catered for the novice computer user as well as the most common needs.
> 
> Additionally, having too many buttons or items could make the toolbar seem intimidating to those aforementioned type of users. I'd prefer things to feel as friendly as possible, but that's definitely a subjective thing.
> 
> Again, the focus for this should really be users with novice computer knowledge, as they are the most common users. This shouldn't be a toolbar of convenience for advanced users, but provide advanced functionality in a way that even users of novice computer knowledge would appreciate and understand.

I'd agree that the program should be catered for novice users out of the box, but it'd be nice if it allowed configuration of the button bar as I mention above, so that advanced users have a very powerful tool at their finger tips for speeding up workflow when doing more advanced things.  Or alternatively, having a global setting "Enable Advanced toolbar for nerds." or something.  :)

sauloco commented:
Is there any ETA for this feature to be implemented?

jp9000 commented:
well it seems I accidentally closed this.

jp9000 commented:
I'll just put this on a branch on the main repo.

ssosina commented:
Hey, when will this feature be released for Windows?

WizardCM commented:
In the next official version. Soon.

ssosina commented:
@WizardCM When will the next official release be?

kkartaltepe commented:
> Soon

MarioMey commented:
Can we know what happened with this PR? I really liked it...

I know that 26.0.2 already has a media controller, but the very usefull control buttons are not there :(.

tt2468 commented:
@MarioMey It's in 26.0.0 afaik, just as a different pr that got merged.

MarioMey commented:
And flip, rotate, scale buttons?

tt2468 commented:
@MarioMey I seem to remember them being left out because they already have simple keyboard shortcuts

tconley commented:
I'd like the media controls in studio mode to show the media that's actually active/playing in program, where right now it's showing whatever is selected in the sources.   Perhaps this needs to live elsewhere?

MarioMey commented:
> I'd like the media controls in studio mode to show the media that's actually active/playing in program, where right now it's showing whatever is selected in the sources. Perhaps this needs to live elsewhere?

Check this: https://obsproject.com/forum/resources/media-controls.1032/
